lib/oli/automation_setup.ex

```diff
@@ -0,0 +1,354 @@
+
+    {:ok, user} =
+      Oli.Accounts.create_author(%{
+        email: generate_email("author"),
+        given_name: "Test",
+        family_name: "Author",
+        password: password,
+        password_confirmation: password,
+        system_role_id: Oli.Accounts.SystemRole.role_id().author,
+        email_confirmed_at: Timex.now()
+      })
+
+    {:ok, user, password}
+  end
+
+  defp has_project(nil) do
+    {:error, "Project not found"}
+  end
+
+  defp has_project(project) do
+    {:ok, project}
+  end
+
+  defp has_no_authors([]) do
+    {:ok}
+  end
+
+  defp has_no_authors(_) do
+    {:error, "Can only delete projects with no authors"}
+  end
+
+  defp can_delete_section(nil) do
+    {:error, "Section not found"}
+  end
+
+  defp can_delete_section(%{:title => "Automation test section"} = section) do
+    {:ok, section}
+  end
+
+  defp can_delete_section(_) do
+    {:error, "Can only delete sections with title 'Automation test section'"}
+  end
+
+  defp no_duplicates(project) do
+    child_project_exists =
+      Repo.exists?(
+        from p in Project,
+          where: p.project_id == ^project.id,
+          select: p
+      )
+
+    if not project.allow_duplication and not child_project_exists do
+      {:ok}
+    else
+      # If we have child projects, or the project allows duplicates, don't let us delete it.
+      {:error, "Project allows duplicates"}
+    end
+  end
+
+  # Tears down a test project
+  # Deletes:
+  # project
+  #   -> publications
+  #       -> publication_resources
+  #   -> project_resources
+  #       -> resources
+  #          -> revisions
+  #
+  # Only works on automation-test projects
+  def teardown_project(slug) do
+    with {:ok, project} <-
+           Repo.one(
+             from p in Project,
+               where: p.slug == ^slug,
+               preload: [
+                 :authors,
+                 :family,
+                 :resources,
+                 :activity_registrations,
+                 :part_component_registrations,
+                 :communities,
+                 :publications
+               ]
+           )
+           |> has_project(),
+         # Only tear down projects where duplicates are not allowed
+         {:ok} <- no_duplicates(project),
+         # Only tear down projects with no authors
+         {:ok} <-
+           has_no_authors(project.authors) do
+      #
+      # Delete all the publications for this project, this will also get the publication_resources
+      Enum.each(project.publications, &Oli.Publishing.delete_publication/1)
+
+      # Delete all the associations between this project and resources
+      Repo.delete_all(from pr in ProjectResource, where: pr.project_id == ^project.id, select: pr)
+
+      # Delete the resources
+      for resource <- project.resources do
+        Repo.delete_all(from r in Oli.Resources.Revision, where: r.resource_id == ^resource.id)
+        Repo.delete(resource)
+      end
+
+      {:ok, _} = Repo.delete(project)
```

Comment:
If we want to raise when this action fails, I think it would be cleaner to use `delete!` instead so that we raise an accurate error.

Reply:
TODO: I think I could get a list of revisions to delete from the PublishedResources before deleting those.